### ISSUE TITLE
Questions about the definition of WS and a possible typo in the paper.

### ISSUE THREAD

**Issue by t-taniai** [joe-siyuan-qiao/WeightStandardization#1]
Thank you for sharing a very nice paper! I have a couple of small questions about the definition of WS.

My first question is about a possible typo in the paper (based on the initial version of https://arxiv.org/abs/1903.10520). I think that the summation \sum_{i=1}^I in the definition of sigma_Wi in Equation 4 should be correctly \sum_{j=1}^I (summation over j, not i), right?

My second question is about the usage of epsilon. In BatchNorm and GroupNorm, they normalize activations by a formulation of x' = (x - Mean(x)) / sqrt( Var(x) + epsilon ), putting the epsilon inside the sqrt. I thought it was because this formulation can avoid zero division in gradients as well as x' itself. In your WS, the normalization of weights is done along some w's axes as w' = (w - Mean(w) / (sqrt( Var(w) ) + epsilon), putting the epsilon outside the sqrt. Unlike the formulation of BN and GN, I think this formulation still has slight possibilities of producing zero division in its gradient computation when Var(w) is exactly 0. Although such a case is very rare, isn't that problematic?

Sorry for bothering you, if I am asking trivial questions or having some misunderstandings.

**Comment by csrhddlam** [comment]
Yes, you are right for both questions. We'll correct the typo in the next version.

Our implementation of std could lead to nan gradients which could affect the whole model, though we didn't observe it in practice. The BN and GN formulation passes zero gradients, which still let the channel rest in peace, but will not contaminate the whole model.

Thanks for your questions.

**Comment by t-taniai** [comment]
Hi, thank you for your quick response. It was good to know that my understandings were right.
I know that what I pointed out won't harm the importance of your work. Thanks for your interesting work.

**Comment by MarcoForte** [comment]
I was getting NaNs after just 276 iterations when fine-tuning ResNet-101 WS GN. 
I followed proposal from @t-taniai  and I haven't had the NaN issue come back yet.
**OLD:**
`std = (weight).view(weight.size(0), -1).std(dim=1).view(-1, 1, 1, 1) + 1e-5  `  
**NEW:**
`std = torch.sqrt(torch.var(weight.view(weight.size(0), -1), dim=1) + 1e-12).view(-1, 1, 1, 1) + 1e-5`

**Comment by bonlime** [comment]
@MarcoForte I've accidentally found your reply and it helped me to fix problems with nan's I've been struggling with for months. Thanks a lot.

**Comment by VCasecnikovs** [comment]
@MarcoForte Hello, unfortunately, your solution didn't help me with NaN. What I found out is that it started to appear in F.conv2D. I tried your std change solution. It didn't help me. Also, I tried doing detach of mean and std calculations. It didn't help me either.
Do you have any thoughts about my problem?
Yours faithfully,
Vadims.

**Comment by bonlime** [comment]
@VCasecnikovs 
are you using mixed precision? i've found that 1e-12 is too small in that case and needs to be set to ~2e-5. 

Adam optimizer also has the problem with eps being outside sqrt. patching it helped me too

**Comment by VCasecnikovs** [comment]
Hello, @bonlime.
Thank you for answering. I've tried 1e-12 and now 2e-5 too. I have just an input explosion. I am using a standard SGD optimizer. The problem is that it explodes on the first forward pass. Doesn't matter if weights are pretrained or not. When I use basic Conv2d, the problem disappears. This is my repo https://github.com/VCasecnikovs/Yet-Another-YOLOv4-Pytorch
Yours faithfully,
Vadims.
